ADR/0029-basic-authentication-secret.md

```diff
@@ -0,0 +1,88 @@
+# 29. Basic Authentication Secret
+
+Date: 2023-08-25
+
+## Status
+
+Proposed
+
+## Context
+
+- As a user, I need a secure way to configure sensitive information/tokens/credentials for various components within the RHTAP environment. This includes providing credentials for accessing source code repositories on GitHub and other authentication needs.
+- As a user, I expect to have a place where I can list already provided credentials and adjust them it if needed.
+
+## Decision
+
+### Persisting Sensitive Information
+
+We propose that the UI captures sensitive information through form input and creates `RemoteSecret` Custom Resources (CRs) to store this data. The `RemoteSecret.data` will be intercepted by a webhook, securely stored, and removed from the `RemoteSecret` CR to ensure that data is not kept in etcd.
+```yaml
+apiVersion: appstudio.redhat.com/v1beta1
+kind: RemoteSecret
+metadata:
+    name: github-com-token
+    namespace: myns
+    labels:
+        appstudio.redhat.com/sp.host: github.com
+spec:
+    secret:
+        type: kubernetes.io/basic-auth
+data:
+  username: Z2VuYQ==
+  password: Z2VuYQ==
+```
+### Accessing Sensitive Information
+
+1. RHTAP components will search for `RemoteSecret` instances based on the `appstudio.redhat.com/sp.host` label.
+2. For the required `RemoteSecret`s, the components will expand the targets using the desired secret name, label, and annotations to access the secret data.
+3. The controller will retrieve `secretData` from permanent storage and deliver it to the desired location.
+4. Once the secret is no longer needed, the components will remove the `RemoteSecret`'s target.
+```yaml
+apiVersion: appstudio.redhat.com/v1beta1
+kind: RemoteSecret
+metadata:
+    name: github-com-token
+    namespace: myns
+    labels:
+        appstudio.redhat.com/sp.host: github.com
+spec:
+    targets:
+    - namespace: myns
+      secret:
+        name: tekton-secret
+        annotations:
+            tekton.dev/git-0: https://github.com
+    secret:
+        type: kubernetes.io/basic-auth
+
+```
+### Fine-Grained Credentials
+
+Users can specify resource names through the UI, which will be added as comma-separated annotation `appstudio.redhat.com/sp.repository` to the `RemoteSecret` CR. RHTAP components will consider these annotations during `RemoteSecret` searches.
+```yaml
+apiVersion: appstudio.redhat.com/v1beta1
+kind: RemoteSecret
+metadata:
+    name: github-com-token
+    namespace: myns
+    annotations:
+        appstudio.redhat.com/sp.repository: "redhat/secret,redhat/opera"
+    labels:
+        appstudio.redhat.com/sp.host: github.com
+spec:
+    secret:
+        type: kubernetes.io/basic-auth
+```
+### Rotation of Tokens/Secrets
+
+Users can rotate tokens/secrets through these steps:
```

Comment:
Will users be able to see what secrets are created so that they know which ones can be revoked externally or overwritten in the UI?

Reply:
This will be achieved by the UI looking for the presence of `RemoteSecret`s as well?

What does it mean to "revoke externally" -- is this via the CLI or would this also be done via the UI somehow?

Reply:
It seems like I might have been confused by the reference point. So it would be external to the system/`RemoteSecret` process... but still in-line via the UI? This ADR is written from the perspective of the remote secret and not from the perspective of AppStudio as a while?

Reply:
I thin we can clarify this section. Maybe something like

```
### Rotation of Tokens/Secrets

It is not possible to delete secrets added to `RemoteSecret`s. Instead, users can revoke or rotate tokens and secrets through these steps:
1. Revoke the secret with at its origination.
2. Generate a new credential with the desired service.
3. Overwrite the secret in RHTAP using the UI.
```